fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/matchers/PlannerBindings.java

```diff
@@ -21,62 +21,147 @@
-     * @param key a matcher with type parameter that is a super class of the class of <code>bindable</code>
-     * @param bindable a bindable object to bind to the key
+     * Retrieve all bindables bound to {@code key} if there is at least one such bindable.
+     * If no bindable is bound to this key, throw a {@link NoSuchElementException}.
+     * @param key a matcher
+     * @param <T> the type of {@link Bindable} that was bound to {@code key}
+     * @return a list of bindable objects bound to the key
+     * @throws NoSuchElementException if no bindable objects are bound to the given key
      */
-    public void put(@Nonnull ExpressionMatcher<? extends Bindable> key, @Nonnull Bindable bindable) {
+    @Nonnull
+    @SuppressWarnings("unchecked")
+    public <T extends Bindable> List<T> getAll(@Nonnull ExpressionMatcher<T> key) {
         if (bindings.containsKey(key)) {
-            throw new RecordCoreException("attempted to add a binding that already exists");
+            return (List<T>) bindings.get(key);
+        }
+        throw new NoSuchElementException("attempted to extract bindable from binding using non-existent key");
+    }
+
+    /**
+     * Combine this set of bindings with the given set of bindings. If both sets of bindings contain a binding with the
+     * same key, the resulting set of bindings will contain the <em>concatenation</em> of the bindables from this set
+     * of bindings and the bindables from the given set of bindings, in that order, providing a predictable ordering
+     * of bindings for planner rules that might need it.
+     * @param other a set of bindings, which may share keys with this set of bindings
+     * @return a new set of bindings that contains the bindings from both both sets of bindings
+     */
+    public PlannerBindings mergeWith(@Nonnull PlannerBindings other) {
+        ListMultimap<ExpressionMatcher<? extends Bindable>, Bindable> combined = MultimapBuilder.hashKeys().arrayListValues().build(this.bindings);
+        combined.putAll(other.bindings);
+        return new PlannerBindings(combined);
+    }
+
+    /**
+     * Build a new set of bindings containing a single binding from the given key to the given bindable.
+     * @param key an expression matcher
+     * @param bindable a bindable object
+     * @return a new set of bindings containign a single binding from {@code key} to {@code bindable}
+     */
+    @Nonnull
+    public static PlannerBindings from(@Nonnull ExpressionMatcher<? extends Bindable> key, @Nonnull Bindable bindable) {
+        return new PlannerBindings(ImmutableListMultimap.of(key, bindable));
+    }
+
+    /**
+     * Return an empty set of bindings.
+     * @return an empty set of bindings
+     */
+    @Nonnull
+    public static PlannerBindings empty() {
+        return EMPTY;
+    }
+
+    /**
+     * Return a new builder.
+     * @return a new builder
+     */
+    public static Builder builder() {
```

Comment:
I think `newBuilder` is more in line with the rest of the code base, though I could be wrong.

It's a little unfortunate that guava went with `builder` and protobuf went with `newBuilder`.

Reply:
I think I prefer `newBuilder`. It's at least what we use for the core things like record stores or record queries.

Reply:
Changed to `newBuilder`.